### PR TITLE
Squash clear cache calls

### PR DESCRIPTION
**Pull request by costinm**
Based on test results, in separate PR I may make it customisable, for now looking to see if it has 
any impact on CPU/memory use.

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v1/discovery.go [hunk 1]
```diff
@@ -481,6 +489,19 @@ func (ds *DiscoveryService) ClearCacheStats(_ *restful.Request, _ *restful.Respo
 // clearCache will clear all envoy caches. Called by service, instance and config handlers.
 // This will impact the performance, since envoy will need to recalculate.
 func (ds *DiscoveryService) clearCache() {
+	clearCacheMutex.Lock()
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/v1/discovery.go]: no need for mutex, this is run on a single event queue

File: pilot/pkg/proxy/envoy/v1/discovery.go [hunk 2]
```diff
@@ -481,6 +489,19 @@ func (ds *DiscoveryService) ClearCacheStats(_ *restful.Request, _ *restful.Respo
...
+	defer clearCacheMutex.Unlock()
+	if time.Since(lastClearCache) < 60*time.Second {
+		if !clearCacheTimerSet {
+			clearCacheTimerSet = true
+			time.AfterFunc(61*time.Second, func() {
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/v1/discovery.go]: isn't this going to block the event queue?

Comment (kyessenov) [reply pilot/pkg/proxy/envoy/v1/discovery.go]: the danger here is the unbound growth of the incoming events. i think we need to run a secondary worker thread to refresh the cache like @rshriram suggested. they only need to share one bit for the "dirtiness" of the cache.

File: istio.VERSION [hunk 3]
```diff
@@ -1,13 +1,13 @@
 # DO NOT EDIT THIS FILE MANUALLY instead use
 # install/updateVersion.sh (see install/README.md)
-export CA_HUB="gcr.io/istio-testing"
-export CA_TAG="replace-by-actual-tag"
-export MIXER_HUB="gcr.io/istio-testing"
-export MIXER_TAG="replace-by-actual-tag"
-export PILOT_HUB="gcr.io/istio-testing"
-export PILOT_TAG="replace-by-actual-tag"
+export CA_HUB="costinm"
```
Comment (ldemailly) [inline istio.VERSION]: please look at your PR/diff before and after submission :-)

File: pilot/pkg/proxy/envoy/v1/discovery.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v1/discovery.go]: Just feel safer with a mutex - with the other changes in multicluster and other optimizations 
we may have multiple callers.

File: istio.VERSION (hunk 3 above)
Comment (costinm) [reply istio.VERSION]: Ah, istio.VERSION again...

File: pilot/pkg/proxy/envoy/v1/discovery.go (hunk 1 above)
Comment (kyessenov) [reply pilot/pkg/proxy/envoy/v1/discovery.go]: how does this mutex work with the cache lock? there are things reading from the cache in parallel, right?

Comment (costinm) [reply pilot/pkg/proxy/envoy/v1/discovery.go]: This mutex only syncs the clearCache calls. 
Additional mutexes or other sync may be needed - we'll discuss it when that work is done, for now
it's useful to stop assuming everything is in one thread... Is there any harm in adding a mutex to clearCache ?

**Comment by kyessenov** [comment]
No, extra mutexes are only dangerous if they introduce dead locks :)

File: pilot/pkg/proxy/envoy/v1/discovery.go [hunk 4]
```diff
@@ -481,6 +499,21 @@ func (ds *DiscoveryService) ClearCacheStats(_ *restful.Request, _ *restful.Respo
 // clearCache will clear all envoy caches. Called by service, instance and config handlers.
 // This will impact the performance, since envoy will need to recalculate.
 func (ds *DiscoveryService) clearCache() {
+	clearCacheMutex.Lock()
+	defer clearCacheMutex.Unlock()
+
+	if time.Since(lastClearCache) < time.Duration(clearCacheTime) * time.Second {
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v1/discovery.go]: greater than ?

Comment (costinm) [reply pilot/pkg/proxy/envoy/v1/discovery.go]: If the time since last clear is less than x seconds - we schedule the event. 
I think it's <

**Comment by costinm** [comment]
ping ?

**Comment by rshriram** [comment]
I would rename this clear cache squash to batchCacheEvictionEvents for clarity.

**Comment by costinm** [comment]
Agreed, will try to do it in the next PR related to this ( to avoid another lgtm/approve/build cycle...)

**Comment by ldemailly** [comment]
the errors in https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3506/e2e-smoke/3037/ seems a bit concerning
/test e2e-smoke

**Pull request merged**